src/IC/Expression.H

```diff
@@ -18,77 +29,61 @@ namespace IC
-                    double loc[3];
-                    loc[0] = x(0); loc[1] = x(1); loc[2] = x(2);
-
-                    field(i,j,k,0) = fparser[0].Eval(loc);
-                    field(i,j,k,1) = fparser[1].Eval(loc);
-                    field(i,j,k,2) = fparser[2].Eval(loc);
-
-//                    Set::Scalar min_value = 100.0;
-//                    for (int i = 0; i < center.size(); i++)
-//                    {
-//                        Set::Scalar value = 0.;
-//                        Set::Scalar norm = (A[i] * (x - center[i])).lpNorm<2>();
-//                        value = 0.5 + 0.5 * std::erf(((x - center[i]).transpose() * A[i] * (x - center[i]) - 1.0) / eps[i] / norm);
-//                        min_value = value < min_value ? value : min_value;
-//                    }
-//                    field(i, j, k) = min_value;
-//                    if (field(i, j, k) < 0.)
-//                        field(i, j, k) = 0.;
-//                    if (field(i, j, k) > 1.)
-//                        field(i, j, k) = 1.;
-                });
+                for (unsigned int n = 0; n < fparser.size(); n++)
+                {
+                    amrex::ParallelFor(bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) {
+                        Set::Vector x = Set::Position(i, j, k, geom[lev], type);
+                        double loc[AMREX_SPACEDIM];
+                        AMREX_D_TERM(loc[0] = x(0);, loc[1] = x(1);, loc[2] = x(2););
+                        field(i,j,k,n) = fparser[n].Eval(loc);
+                    });
+                }
             }
             a_field[lev]->FillBoundary();
         };
 
     private:
-    //
-    // TODO: this part still needs to be updated. Cannot yet handle arbitrary number of grains...
-    //
-        std::array<FunctionParserAD,3> fparser;
-        //	int dim = AMREX_SPACEDIM;
-        //	amrex::Vector<Set::Vector> center;
-        //	amrex::Vector<Set::Scalar> eps;
-        //	amrex::Vector<Set::Matrix> A;
+        std::vector<FunctionParserAD> fparser;
 
     public:
         static void Parse(Expression &value, IO::ParmParse &pp)
         {
-            std::array<char *,3> names = {"region1","region2","region3"};
-            for (int i = 0; i < value.fparser.size(); i++)
+            for (int i = 0; true; i++)
             {
                 std::string func = "0.0";
-                pp.query(names[i],func);
+                std::string name = "region" + std::to_string(i);
+                
+                if (!pp.contains(name.data())) break;
+                pp.query(name.data(),func);
+
+                FunctionParserAD fparser;
 
-                value.fparser[i].AddVariable("x");
-                value.fparser[i].AddVariable("y");
-                value.fparser[i].AddVariable("z");
-                if (value.fparser[i].Parse(func.c_str(), "x,y,z") != -1)
+                fparser.AddVariable("x");
+                fparser.AddVariable("y");
+                fparser.AddVariable("z");
```

Comment:
Do these lines work in 2D?

Reply:
Yes, it will work in 2D, it just won't do anything.